### ISSUE TITLE
Database issues

### ISSUE THREAD

**Issue by hazmat345** [beer-garden/beer-garden#84]
- [x] Database connections are timing out
- [x] Requests page takes a long time

**Comment by hazmat345** [comment]
These are biting us:
https://groups.google.com/forum/#!topic/mongodb-user/O-fMpLuzulY
https://jira.mongodb.org/browse/SERVER-12869

As far as I can tell it doesn't seem to be possible to create a negative existence index that can count using _only_ the index. Counting needs to 'confirm' the nullity by loading each document.

We could maybe work around this by introducing some sort of `is_root` flag and indexing that.

**Comment by loganasherjones** [comment]
I think, short of MongoDB adding this new capability, we should add the `is_parent` flag to the request. I don't see this requirement going away for us.  As far as I can tell, the only other solution is to actually nest the documents which isn't that bad, but I believe those changes will be bigger than I'm willing to put into a patch release.

**Comment by hazmat345** [comment]
To close the loop on this:

### Connection Timeouts
Database connections were timing out because of our startup mongo connection checking. We lower the timeout to 1s (from 30s) so that we can control the retry timings. We were reusing the connection with those timeouts for all operations, which was bad. This was addressed by using a separate connection for the initial aliveness check. When the check succeeds we close that connection and open another one with the correct timeouts.

### Request Page Slowness
The requests page was taking a long time because the mongo indexes we were using weren't applicable to the queries we were running. There were a couple of things happening:

- The index on `parent` was the most important, and it wasn't working. Because of how mongo indexes handle the null case (the value is null vs the key not existing) trying to index on a negative existence check doesn't work - mongo needs to actually inspect all the documents to ensure they match.
- Another issue was that index intersection doesn't apply when sorting and querying depend on completely disjoint indexes. In practice this resulted in mongo always using the sorting field's index _only_, which meant that _all_ documents needed to be inspected.

To address this we added an additional `has_parent` boolean field to the request model. We then modified the indexes for the fields displayed in the request index page to be compound with the `has_parent` field being the primary. This will allow mongo to chop out all child requests right away while still maintaining sorting speed.

This should result in a marked improvement in the non-filtering cases as they should result in covered queries. It should also help in the filtered cases. For filters with an exact match it's possible to have a covered query if the sorting is also on that field (i.e. it should be fast to filter on `status` if the sorting is also `status`). There should also be some improvement for non-covered cases since mongo will only need to examine parent requests instead of the entire collection.

**Comment by hazmat345** [comment]
These are the mongo commands to update the request collection manually:

```
db.request.update( { parent: { $not: { $eq: null } } }, { $set : { is_parent: false } }, { multi: true } )
db.request.update( { parent: null }, { $set : { is_parent: true } }, { multi: true } )
```

**Comment by hazmat345** [comment]
Fixed by beer-garden release [2.3.7](https://github.com/beer-garden/beer-garden/releases/tag/2.3.7)
